### DISCUSSION TITLE
Breadcrumb feature (can eglot support headerline like lsp-mode does?)

### DISCUSSION THREAD

**Discussion by haoisli9** [joaotavora/eglot#988] [Q&A] [answered]
I really like eglot but it has not the headerline support which i always used in lsp-mode to get the node structure or location. it is very useful when I edit yaml file. the node entry is so deep and i always forgot where i am.
lsp-mode is too heavy and slow. so i change to eglot. but i can not find this feature in eglot. so can i get this in the futher?

and, how can i config yaml server location? my yaml server location is not in the path, so i want to config it by myself. how can i do?

thanks.

**Comment by joaotavora** [comment, marked as answer]
> I really like eglot but it has not the headerline support which i always used in lsp-mode to get the node structure or location.

Alright, I more or less see what you mean.  In lsp-mode you use the headerline to give you some kind of nesting path of the current point.  But I need you or someone else to guide me in the way of the LSP feature that provides this.  Then we can organize the design of this feature.  When you have found this, make a **new** _feature request discussion_ with the collected information, and some screenshots of how you think the feature should look like.

> and, how can i config yaml server location? my yaml server location is not in the path, so i want to config it by myself. how can i do?

Use `eglot-server-programs`.  Learn how to read docstrings of variables to set the values of variables in Emacs, either via the "customize" interface, which some find more user-friendly or via elisp directly.  Do learn this, it'll pay off.  But in the meantime, I can help you out with this request.


Since `eglot-server-programs` is an alist we can use `alist-get` to access its elements.  If we just use `alist-get` we'll be reading the entry, but we can add `setf` to transform that access into a write.  So armed with that knowledge and the first few lines of the docstring of `eglot-server-programs`, we can craft the following line:
```
(setf (alist-get 'yaml-mode eglot-server-programs) "my/special/path/to/yaml/server") 
```

Put this somewhere in your init file (`~/.emacs` or `~/emacs.d/init.el`).  It'll change the entry for `yaml-mode` in your `eglot-server-programs`.

**Comment by haoisli9** [comment]
i can not attach image here. would you please give me another email address so I can send you the image that shows what lsp-mode does.
thanks.

BTW, i change the setting to (setf (car (alist-get 'yaml-mode eglot-server-programs))  "my/special/path/to/yaml/server") it seems works for me. 
and, would you add some lsp-server installation function? lsp-mode support server installation and easy for the newer.

**Reply by joaotavora** [reply]
> would you add some lsp-server installation function

No. Eglot is middleware to talk to servers. It tries to do one thing and do it well.  It's supposed to be small.  Also installing other software on the users machine has practical, ethical and legal considerations. You or someone else can build some other tool on top of Eglot. 
> BTW, i change the setting

That's very odd, but who knows :shrug: 

> i can not attach image here. would you please give me another email address so I can send you the image that shows what lsp-mode does.

No. You can attach images where.  Just drag and drop.  But it's not very important. I think you are requesting a "breadcrumb" path in the header-line, right?

The most important thing is not the image, is to find the part of the LSP spec (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/) where they talk about how to ask the server for the information about the breadcrumb.

**Comment by haoisli9** [comment]
Yes you are right, it's breadcrumb shows in headerline.




***@***.***
 
From: João Távora
Date: 2022-07-05 14:59
To: joaotavora/eglot
CC: haoisli9; Author
Subject: Re: [joaotavora/eglot] can eglot support headerline like lsp-mode does? (Discussion #988)
would you add some lsp-server installation function
No. Eglot is middleware to talk to servers. It tries to do one thing and do it well. It's supposed to be small. Also installing other software on the users machine has practical, ethical and legal considerations. You or someone else can build some other tool on top of Eglot.
BTW, i change the setting
That's very odd, but who knows 
i can not attach image here. would you please give me another email address so I can send you the image that shows what lsp-mode does.
No. You can attach images where. Just drag and drop. But it's not very important. I think you are requesting a "breadcrumb" path in the header-line, right?
The most important thing is not the image, is to find the part of the LSP spec (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/) where they talk about how to ask the server for the information about the breadcrumb.

**Comment by tylerjl** [comment]
Hello hello! My apologies if I'm breaching etiquitte by resurrecting this old thread, but I was looking for the same feature (breadcrumb support in a header-like UI element) and landed here. For completeness, here's what a breadcrumb might look like for a Haskell file - first, while hovering over a data/sum type:

![image](https://user-images.githubusercontent.com/1733846/192888388-d2eb7631-7ca0-44cb-bede-bb8e23b16397.png)

...and while hovering the cursor while inside `main`:

![image](https://user-images.githubusercontent.com/1733846/192888501-a25426ae-a218-4274-9a13-f992631278a9.png)

Per @joaotavora's request, [the relevant portion of th LSP spec appears to be `DocumentSymbol`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol). I popped into some other LSP clients and they all seem to rely on `DocumentSymbol[]` to construct the breadcrumb because it returns a hierarchy, while `SymbolInformation[]` does not. Although I provided the example UI here, I am _very_ much an elisp noob and have _no_ idea what the "correct" mechanism should be for actually displaying breadcrumb elements (though I'm willing to help to the extent I can :smile:). 

Thanks!

**Reply by joaotavora** [reply]
> equest, [the relevant portion of th LSP spec appears to be DocumentSymbol](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_documentSymbol). 

Yes, and it turns out Eglot already "speaks" it, and uses it to setup an imenu index.

Eglot recently had an imenu-related "bug" fixed.  This means that if you try `M-x imenu` anywhere in an Eglot-managed buffer, you'll probably find this "breadcrumb" information displayed -- only differently.

What we need right now, is an kind of `put-a-breadcrumb-in-headerline-from-imenu-index.el` library, that doesn't have anything to do with Eglot. But it'll do the right thing since Eglot supplies an imenu index.  One library that is almost that is the `which-func.el` library.  

Try `M-x which-func-mode` and look at the mode-line.  It'll tell you the current function.  Again, based on imenu.  I guess you need to write something similar but for the headerline and which displays the whole path.  And probably prettier.  So if you're an Elisp noob try to find someone with elisp skills and refer them here.

**Reply by tylerjl** [reply]
Thank you @joaotavora, that's very helpful. I may tinker here and there with a solution, but having literally never written an elisp library, any readers here should _not_ assume that I'll be providing something any time soon. But hopefully this discussion is a useful checkpoint if any experienced library authors want to string together that imenu+breadcrumb combination.

**Comment by haoisli9** [comment]
Yes you are right, it is breadcrumb.
Can eglot support breadcrumb feature? or does eglot have the plan ?




***@***.***
 
From: João Távora
Date: 2022-07-05 14:59
To: joaotavora/eglot
CC: haoisli9; Author
Subject: Re: [joaotavora/eglot] can eglot support headerline like lsp-mode does? (Discussion #988)
would you add some lsp-server installation function
No. Eglot is middleware to talk to servers. It tries to do one thing and do it well. It's supposed to be small. Also installing other software on the users machine has practical, ethical and legal considerations. You or someone else can build some other tool on top of Eglot.
BTW, i change the setting
That's very odd, but who knows 
i can not attach image here. would you please give me another email address so I can send you the image that shows what lsp-mode does.
No. You can attach images where. Just drag and drop. But it's not very important. I think you are requesting a "breadcrumb" path in the header-line, right?
The most important thing is not the image, is to find the part of the LSP spec (https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/) where they talk about how to ask the server for the information about the breadcrumb.

**Reply by joaotavora** [reply]
> Yes you are right, it is breadcrumb.
> Can eglot support breadcrumb feature? or does eglot have the plan ?

It's not something for eglot to support, rather for Emacs's imenu.el to support. 

I've filed an Emacs feature request for this: https://debbugs.gnu.org/cgi/bugreport.cgi?bug=58431

Suggest to continue the discussion there.

**Comment by gdanov** [comment]
**update** : don't use this hacky implementation, see at the bottom for more robust implementation.


I'm also on the opinion that this should be add-on and not core eglot function. 

I've put together something temporary that gets the job done for me. It piggybacks on `which-function-mode` which on it's side uses imenu's eglot integration. The solution is probably more of a generic imenu-driven breadcrumb, not eglot specific.

Here is screenshot and current version of the code

<img width="843" alt="Screenshot 2022-12-02 at 11 31 12" src="https://user-images.githubusercontent.com/957113/205272866-c697e706-44ff-4b8b-ba30-95c2366b9e0e.png">

```
(setq
 which-func-imenu-joiner-function
 (lambda (x)
	 (let* ((otype (car x))
					(loc-sym (car (split-string (car (last x)) " ")))
					(path (mapcan
								 (lambda (pel)
									 (split-string pel "::"))
								 (seq-map
									(lambda (pel)
										(car (split-string pel " ")))
									(butlast (cdr x)))))
					(result (seq-filter
									 (lambda (pel)
										 (and
											(<= (length (split-string pel "::")) 1)
											;;(<= (length (split-string pel "\\.")) 1)
											(seq-reduce (lambda (o v)
																		(and o (not (string= pel v)))) (list "callback") t)))
									 path)))

		 (when (seq-reduce (lambda (o v) (or o (string= v otype))) (list "Method" "Constructor" "Function") nil)
			 (setq result (append result (list loc-sym))))

		 ;;(message "<< %s %s" x (list (list loc-sym ":" otype) (list (seq-map #'list path))))
		 ;;(message ">>> %s" result)

		 (setq header-line-format (mapconcat #'identity result " » "))
		 ;; return
		 "")))
```

**Reply by tylerjl** [reply]
This is really cool @gdanov. After learning more about the underlying systems that eglot sits atop, your and @joaotavora's responses make a lot of sense. Part of the entire _reason_ why eglot and its tight integration is appealing is that existing systems can be re-used, like imenu.

I'd love to have something like this and it does seem well-suited as its own thing. I don't know whether that's its own, small package or in core but I can certainly imagine something like a `imenu-breadcrumb` minor mode.

**Reply by gdanov** [reply]
> something like a `imenu-breadcrumb` minor mode.

yes, exactly my point. BTW my piggyback solution is pretty dirty and probably bad design, but the best I can do at the moment.

**Comment by joaotavora** [comment]
imenu-breadcrumb minor mode is a very good name.  Suggest you submit this
suggestion
to the existing https://debbugs.gnu.org/cgi/bugreport.cgi?bug=58431 I
created.

Also suggest that discussion continue there.  And that François-René's
solution
is attached there.

**Comment by haoisli9** [comment]
Seems excellent!

But how can I do to let it work without elgot? I had enabled which-func-mode. but seems it does not work? Does it depend on elgot? or maybe I should add something?

Thanks.



***@***.***
 
From: gdanov
Date: 2022-12-02 18:34
To: joaotavora/eglot
CC: haoisli9; Author
Subject: Re: [joaotavora/eglot] Breadcrumb feature (can eglot support headerline like lsp-mode does?) (Discussion #988)
I'm also on the opinion that this should be add-on and not core eglot function.
I've put together something temporary that gets the job done for me. It piggybacks on which-function-mode which on it's side uses imenu's eglot integration. The solution is probably more of a generic imenu-driven breadcrumb, not eglot specific.
Here is screenshot and current version of the code
(setq
 which-func-imenu-joiner-function
 (lambda (x)
	 (let* ((otype (car x))
					(loc-sym (car (split-string (car (last x)) " ")))
					(path (mapcan
								 (lambda (pel)
									 (split-string pel "::"))
								 (seq-map
									(lambda (pel)
										(car (split-string pel " ")))
									(butlast (cdr x)))))
					(result (seq-filter
									 (lambda (pel)
										 (and
											(<= (length (split-string pel "::")) 1)
											;;(<= (length (split-string pel "\\.")) 1)
											(seq-reduce (lambda (o v)
																		(and o (not (string= pel v)))) (list "callback") t)))
									 path)))
		 (when (seq-reduce (lambda (o v) (or o (string= v otype))) (list "Method" "Constructor" "Function") nil)
			 (setq result (append result (list loc-sym))))
		 
		 ;;(message "<< %s %s" x (list (list loc-sym ":" otype) (list (seq-map #'list path))))
		 ;;(message ">>> %s" result)
		 (setq header-line-format (mapconcat #'identity result " » "))
		 ;; return
		 "")))

**Comment by the42** [comment]
Are there plans to integrate that into stock emacs/eglot?

As it stands right now, copy/paste the code from @gdanov  into init.el is what we get, right?

**Comment by joaotavora** [comment]
See this: https://github.com/joaotavora/breadcrumb

**Reply by gdanov** [reply]
sounds like much better solution than my hack
